### PR TITLE
Fix for duplicate dependencies in ASR

### PR DESCRIPTION
**Pull request by czgdp1807**
Closes https://github.com/lfortran/lfortran/issues/1511

### CODE REVIEW

**Comment by czgdp1807** [comment]
I am yet to pick some changes from #1693.

**Comment by czgdp1807** [comment]
I am getting the following error repeatedly on the CI,

```zsh
[NbConvertApp] Writing 16820 bytes to Demo2_out.nbconvert.ipynb
cat: Demo1_out.ipynb: No such file or directory
Traceback (most recent call last):
  File "ci/build_tmp.xsh", line 78, in <module>
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/built_ins.py", line 196, in subproc_captured_hiddenobject
    return xonsh.procs.specs.run_subproc(cmds, captured="hiddenobject", envs=envs)
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/procs/specs.py", line 903, in run_subproc
    return _run_specs(specs, cmds)
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/procs/specs.py", line 936, in _run_specs
    command.end()
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/procs/pipelines.py", line 458, in end
    self._end(tee_output=tee_output)
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/procs/pipelines.py", line 477, in _end
    self._raise_subproc_error()
  File "/home/runner/micromamba-root/envs/lf/lib/python3.10/site-packages/xonsh/procs/pipelines.py", line 599, in _raise_subproc_error
    raise subprocess.CalledProcessError(rtn, spec.args, output=self.output)
subprocess.CalledProcessError: Command '['cat', 'Demo1_out.ipynb']' returned non-zero exit status 1.
ci/build_tmp.xsh:49:$lfortran_version=$(cat version).strip()
ci/build_tmp.xsh:50:$dest="lfortran-" + $lfortran_version
ci/build_tmp.xsh:51:bash ci/create_source_tarball0.sh
ci/build_tmp.xsh:52:tar xzf dist/lfortran-$lfortran_version.tar.gz
ci/build_tmp.xsh:53:cd lfortran-$lfortran_version
ci/build_tmp.xsh:72:jupyter kernelspec list --json
ci/build_tmp.xsh:75:cd share/lfortran/nb
Error: Process completed with exit code 1.
```

**Comment by czgdp1807** [comment]
I investigated. `nbconvert=7.3.1` names the generated file as `Demo1_out.nbconvert.ipynb` and `nbconvert<7.3.1` names it as `Demo1_out.ipynb`. `nbconvert 7.3.1` is getting installed on the CI but the build script is written according to `nbconvert < 7.3.1`.

File: src/libasr/asr_verify.cpp [hunk 1]
```diff
@@ -242,6 +242,18 @@ class VerifyVisitor : public BaseWalkVisitor<VerifyVisitor>
...
+        // Check if any dependency is duplicated
+        // in the dependency list of the function
+        std::set<std::string> dependencies_set;
+        for( size_t i = 0; i < x.n_dependencies; i++ ) {
+            std::string found_dep = x.m_dependencies[i];
+            require(dependencies_set.find(found_dep) == dependencies_set.end(),
+                    "Symbol " + found_dep + " is duplicated in the dependency "
+                    "list of " + std::string(x.m_name));
+            dependencies_set.insert(found_dep);
+        }
```
Comment (certik) [inline src/libasr/asr_verify.cpp:255]: This logic is duplicated 4 times, I would refactor it into a dedicated function to be used like `require(!contains_duplicates(x.n_dependencies, x.m_dependencies), ...)`

File: src/libasr/asdl_cpp.py [hunk 2]
```diff
@@ -2178,7 +2178,10 @@ def visitProduct(self, prod, name):
                     if field.type == "identifier":
                         self.emit('{', 2)
                         self.emit('uint64_t n = self().read_int64();', 3)
-                        self.emit("Vec<char*> v;", 3)
+                        if field.name == "dependencies":
+                            self.emit("DependenciesVec v_%s;" % (field.name), 3)
+                        else:
+                            self.emit("Vec<char*> v_%s;" % (field.name), 3)
```
Comment (certik) [inline src/libasr/asdl_cpp.py:2184]: Why is this needed? The serialized code should have unique dependencies already.

**Review by certik — COMMENTED** [review]
I have couple comments.

File: src/libasr/containers.h [hunk 3]
```diff
@@ -135,6 +135,87 @@ struct Vec {
...
+        if( !reserved ) {
+            reserve(al, 0);
+        }
+
+        if( !present(x) ) {
+            Vec<char*>::push_back(al, x);
+        }
+    }
+
+};
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/libasr/containers.h:217]: Well you can’t compare `char*` directly by `==`. You need to call `strcmp`. Also you can’t call `strcmp` over templated type `T*`.

File: src/libasr/asdl_cpp.py (hunk 2 above)
Comment (czgdp1807) [reply src/libasr/asdl_cpp.py:2184]: Yes. I was compelled to use it here. I didn’t want to. I will see it we can revert to `Vec`.

File: src/libasr/containers.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (certik) [reply src/libasr/containers.h:217]: You have to specialize it and implement using `strcmp` for `char*`.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/libasr/containers.h:217]: Do you mean to shift this push_back to Vec and remove it from here? Well do we always want to keep unique elements in a Vec<char*> class? Is repetition always prohibited in a vector of strings?

[the remark this answers is not in the document]
Comment (certik) [reply src/libasr/containers.h:217]: > Do you mean to shift this push_back to Vec and remove it from here?

I would call it "push_back_unique". It's a good point to enforce uniqueness via a type like you did in this PR. In that case, I would still implement "push_back_unique" in Vec, and then subclass Vec like you do, and change push_back to just call push_back_unique, so the whole implementation will be less than 10 lines. That will be clear and maintainable.

> Well do we always want to keep unique elements in a Vec<char*> class?

Not in general.

> Is repetition always prohibited in a vector of strings?

I would not assume it is always prohibited.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/libasr/containers.h:224]: Done.

**Comment by czgdp1807** [comment]
Weird stuff - https://github.com/lfortran/lfortran/actions/runs/4665109822/jobs/8258088652?pr=1532. 

The following completes and then Github cancels the operation. 🥲.

```yml
- name: Test built lfortran.wasm
        shell: bash -l {0}
        run: |
            set -ex
            source $HOME/ext/emsdk/emsdk_env.sh # Activate Emscripten
            node src/lfortran/tests/test_lfortran.js
```

**Comment by certik** [comment]
I rerun it, let's see if it fixes it.

**Comment by Shaikh-Ubaid** [comment]
I am working on fixing the CI failure here https://github.com/Shaikh-Ubaid/lfortran/pull/12.

**Comment by Shaikh-Ubaid** [comment]
>I am working on fixing the CI failure here https://github.com/Shaikh-Ubaid/lfortran/pull/12.

I think I found the fix. Cleaning history and submitting.

**Comment by czgdp1807** [comment]
This got closed.

File: src/libasr/containers.h [hunk 4]
```diff
@@ -135,6 +178,51 @@ struct Vec {
 static_assert(std::is_standard_layout<Vec<int>>::value);
 static_assert(std::is_trivial<Vec<int>>::value);
 
+struct DependenciesVec: Vec<char*> {
```
Comment (certik) [inline src/libasr/containers.h:181]: Add documentation: The motivation for this class is to emulate the `std::set` API, so that `DependenciesVec` acts as a drop in replacement. A better name would be something like `SetChar`.

**Review by certik — APPROVED** [review]
I think this looks good now, thanks!

**Comment by certik** [comment]
@czgdp1807 thanks for this, looks very good.
